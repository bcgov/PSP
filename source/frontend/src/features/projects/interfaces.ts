export interface IProjectFilter {
  projectRegionCode: string;
  projectStatusCode: string;
  projectName: string;
  projectNumber: string;
}

export interface IProjectSearchBy {
  projectNumber: string;
  projectName: string;
}
