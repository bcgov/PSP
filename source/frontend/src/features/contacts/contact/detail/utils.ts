import { ContactInfoField } from 'features/contacts/interfaces';
import { Dictionary } from 'interfaces/Dictionary';
import { IContactAddress, IContactMethod } from 'interfaces/IContact';

import { AddressTypes } from './../../../../constants/addressTypes';

// the order of this array corresponds to the expected display order
const addressSortOrder = [AddressTypes.Mailing, AddressTypes.Residential, AddressTypes.Billing];

export const sortAddresses = (a1: IContactAddress, a2: IContactAddress) => {
  if (a2.addressType.id === a1.addressType.id) {
    return 0;
  }
  const a2Index = addressSortOrder.indexOf(a2.addressType.id as AddressTypes);
  if (a2Index !== -1 && a2Index < addressSortOrder.indexOf(a1.addressType.id as AddressTypes)) {
    return 1;
  }

  return -1;
};

export function getContactInfo(
  validTypes: Dictionary<string>,
  contactMethods?: IContactMethod[],
): ContactInfoField[] {
  if (contactMethods === undefined) {
    return [];
  }
  // Get only the valid types
  let filteredFields = contactMethods.reduce(
    (accumulator: ContactInfoField[], method: IContactMethod) => {
      if (Object.keys(validTypes).includes(method.contactMethodType.id)) {
        accumulator.push({
          info: method.value,
          label: validTypes[method.contactMethodType.id],
        });
      }
      return accumulator;
    },
    [],
  );

  // Sort according to the dictionary order
  return filteredFields.sort((a, b) => {
    return Object.values(validTypes).indexOf(a.label) - Object.values(validTypes).indexOf(b.label);
  });
}

export const fakeAddresses = [
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'BILLING',
      description: 'Billing address',
      isDisabled: false,
    },
    streetAddress1: 'Billing address',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6A 5G7',
  },
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'MAILING',
      description: 'Mailing address',
      isDisabled: false,
    },
    streetAddress1: 'Mailing address',
    streetAddress2: 'Living in a van',
    streetAddress3: 'Down by the River',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6Z 5G7',
  },
  {
    id: 5,
    rowVersion: 1,
    addressType: {
      id: 'RESIDNT',
      description: 'Property address',
      isDisabled: false,
    },
    streetAddress1: 'Property address',
    streetAddress2: '',
    streetAddress3: '',
    municipality: 'vic',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V9A 0H6',
  },
];
