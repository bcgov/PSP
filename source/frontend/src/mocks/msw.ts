import {
  getMockAddresses,
  getMockLandChars,
  getMockLegalDescriptions,
  getMockValues,
} from 'mocks/bcAssessmentMock';
// src/mocks.js
// 1. Import the library.
import { rest, setupWorker } from 'msw';

import { getMockDescription, getMockSales } from './bcAssessmentMock';

// 2. Describe network behavior with request handlers.
export const worker = setupWorker(
  rest.get('https://delivery.apps.gov.bc.ca/ext/sgw/geo.bca', (req, res, ctx) => {
    if (req.url.search.includes('BCA_FOLIO_ADDRESSES_SV')) {
      return res(ctx.delay(500), ctx.status(200), ctx.json(getMockAddresses()));
    } else if (req.url.search.includes('BCA_FOLIO_LEGAL_DESCRIPTS_SV')) {
      return res(ctx.delay(500), ctx.status(200), ctx.json(getMockLegalDescriptions()));
    } else if (req.url.search.includes('BCA_FOLIO_GNRL_PROP_VALUES_SV')) {
      return res(ctx.delay(500), ctx.status(200), ctx.json(getMockValues()));
    } else if (req.url.search.includes('BCA_FOLIO_SALES_SV')) {
      return res(ctx.delay(500), ctx.status(200), ctx.json(getMockSales()));
    } else if (req.url.search.includes('BCA_FOLIO_DESCRIPTIONS_SV')) {
      return res(ctx.delay(500), ctx.status(200), ctx.json(getMockDescription()));
    } else if (req.url.search.includes('BCA_FOLIO_LAND_CHARS_SV')) {
      return res(ctx.delay(500), ctx.status(200), ctx.json(getMockLandChars()));
    }
  }),
);
