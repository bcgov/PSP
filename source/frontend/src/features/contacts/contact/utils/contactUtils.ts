import { CountryCodes } from 'constants/index';
import { AddressField } from 'features/contacts/interfaces';
import { validateYupSchema, yupToFormErrors } from 'formik';
import { IEditableOrganizationForm, IEditablePersonForm } from 'interfaces/editable-contact';
import isEmpty from 'lodash/isEmpty';

import {
  hasAddress,
  hasEmail,
  hasPhoneNumber,
  OrganizationValidationSchema,
  PersonValidationSchema,
} from '../create/validation';
import { sortAddresses } from '../detail/utils';
import { IContactAddress } from './../../../../interfaces/IContact';

export const onValidateOrganization = (
  values: IEditableOrganizationForm,
  otherCountryId?: string,
) => {
  const errors = {} as any;
  try {
    // combine yup schema validation with custom rules
    if (!hasEmail(values) && !hasPhoneNumber(values) && !hasAddress(values)) {
      errors.needsContactMethod =
        'Contacts must have a minimum of one method of contact to be saved. (ex: email,phone or address)';
    }
    validateYupSchema(values, OrganizationValidationSchema, true, {
      otherCountry: otherCountryId,
    });
    return errors;
  } catch (err) {
    return { ...errors, ...yupToFormErrors(err) };
  }
};

export const onValidatePerson = (values: IEditablePersonForm, otherCountryId?: string) => {
  const errors = {} as any;
  try {
    // combine yup schema validation with custom rules
    if (!hasEmail(values) && !hasPhoneNumber(values) && !hasAddress(values)) {
      errors.needsContactMethod =
        'Contacts must have a minimum of one method of contact to be saved. (ex: email,phone or address)';
    }
    validateYupSchema(values, PersonValidationSchema, true, { otherCountry: otherCountryId });

    return errors;
  } catch (err) {
    return { ...errors, ...yupToFormErrors(err) };
  }
};

export const toAddressFields = (addresses: IContactAddress[]) => {
  return addresses
    .sort(sortAddresses)
    .reduce((accumulator: AddressField[], value: IContactAddress) => {
      accumulator.push({
        label: value.addressType.description || '',
        streetAddress1: value.streetAddress1,
        streetAddress2: value.streetAddress2,
        streetAddress3: value.streetAddress3,
        municipalityAndProvince: [
          value.municipality !== undefined ? value.municipality : '',
          value.province?.provinceStateCode ?? '',
        ]
          .filter(x => !isEmpty(x))
          .join(' '),
        country:
          value.country?.countryCode === CountryCodes.Other
            ? value.countryOther ?? ''
            : value.country?.description,
        postal: value.postal,
      });
      return accumulator;
    }, []);
};
