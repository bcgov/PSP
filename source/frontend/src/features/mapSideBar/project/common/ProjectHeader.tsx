import { Col } from 'react-bootstrap';

import AuditSection from '@/components/common/HeaderField/AuditSection';
import { HeaderField } from '@/components/common/HeaderField/HeaderField';
import StatusField from '@/components/common/HeaderField/StatusField';
import { StyledFiller, StyledRow } from '@/components/common/HeaderField/styles';
import { ApiGen_Concepts_Project } from '@/models/api/generated/ApiGen_Concepts_Project';
import { exists } from '@/utils';

import { StyledLeftHeaderPane } from '../../shared/header/styles';

export interface IProjectHeaderProps {
  project?: ApiGen_Concepts_Project;
}

const ProjectHeader: React.FunctionComponent<React.PropsWithChildren<IProjectHeaderProps>> = ({
  project,
}) => {
  return (
    <StyledRow className="no-gutters">
      <StyledLeftHeaderPane xl="7" xs="12">
        <HeaderField label="Project:" labelWidth={{ xs: 3 }} contentWidth={{ xs: 9 }}>
          {project?.code} {project?.description}
        </HeaderField>
        <HeaderField label="MoTI region:" labelWidth={{ xs: 3 }} contentWidth={{ xs: 9 }}>
          {project?.regionCode?.description}
        </HeaderField>
      </StyledLeftHeaderPane>
      <Col>
        <StyledFiller>
          <AuditSection baseAudit={project} />
          {exists(project?.projectStatusTypeCode) && (
            <StatusField statusCodeType={project?.projectStatusTypeCode} />
          )}
        </StyledFiller>
      </Col>
    </StyledRow>
  );
};

export default ProjectHeader;
