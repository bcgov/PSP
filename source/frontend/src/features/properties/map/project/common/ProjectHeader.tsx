import { UserNameTooltip } from 'components/common/UserNameTooltip';
import { HeaderField } from 'features/mapSideBar/tabs/HeaderField';
import { Api_Project } from 'models/api/Project';
import { Col, Row } from 'react-bootstrap';
import styled from 'styled-components';
import { prettyFormatDate } from 'utils';

export interface IProjectHeaderProps {
  project?: Api_Project;
}

const ProjectHeader: React.FunctionComponent<React.PropsWithChildren<IProjectHeaderProps>> = ({
  project,
}) => {
  return (
    <StyledRow className="no-gutters">
      <Col xs="7">
        <Row className="no-gutters">
          <Col>
            <HeaderField label="Project:" labelWidth="3" contentWidth="9">
              {project?.code} {project?.description}
            </HeaderField>
          </Col>
        </Row>
        <Row className="no-gutters">
          <Col>
            <HeaderField label="MoTI Region:" labelWidth="3" contentWidth="9">
              {project?.regionCode?.description}
            </HeaderField>
          </Col>
        </Row>
      </Col>
      <Col xs="5">
        <Row className="no-gutters">
          <Col className="text-right">
            <StyleSmallText>
              Created: <strong>{prettyFormatDate(project?.appCreateTimestamp)}</strong> by{' '}
              <UserNameTooltip
                userName={project?.appCreateUserid}
                userGuid={project?.appCreateUserGuid}
              />
            </StyleSmallText>
          </Col>
        </Row>
        <Row className="no-gutters">
          <Col className="text-right">
            <StyleSmallText>
              Last updated: <strong>{prettyFormatDate(project?.appLastUpdateTimestamp)}</strong> by{' '}
              <UserNameTooltip
                userName={project?.appLastUpdateUserid}
                userGuid={project?.appLastUpdateUserGuid}
              />
            </StyleSmallText>
          </Col>
        </Row>
        <Row className="no-gutters">
          <Col>
            <HeaderField className="justify-content-end" label="Status:">
              {project?.projectStatusTypeCode?.description}
            </HeaderField>
          </Col>
        </Row>
      </Col>
    </StyledRow>
  );
};

export default ProjectHeader;

const StyledRow = styled(Row)`
  margin-top: 0.5rem;
  margin-bottom: 1.5rem;
  border-bottom-style: solid;
  border-bottom-color: grey;
  border-bottom-width: 0.1rem;
`;

const StyleSmallText = styled.span`
  font-size: 0.87em;
  line-height: 1.9;
`;
