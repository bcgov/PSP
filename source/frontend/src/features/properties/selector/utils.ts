import { IMapProperty } from 'features/properties/selector/models';
import { compact } from 'lodash';
import { Api_PropertyFile } from 'models/api/PropertyFile';
import { formatApiAddress, pidFormatter } from 'utils';

import { Api_Geometry, Api_Property } from './../../../models/api/Property';

export enum NameSourceType {
  PID = 'PID',
  PIN = 'PIN',
  PLAN = 'Plan #',
  LOCATION = 'Location',
  NAME = 'Descriptive Name',
  NONE = 'n/a',
  ADDRESS = 'Address',
}

interface PropertyName {
  label: NameSourceType;
  value: string;
}

export const getPropertyName = (property: IMapProperty): PropertyName => {
  if (property.pid !== undefined && property.pid?.toString().length > 0 && property.pid !== '0') {
    return { label: NameSourceType.PID, value: pidFormatter(property.pid.toString()) };
  } else if (
    property.pin !== undefined &&
    property.pin?.toString()?.length > 0 &&
    property.pin !== '0'
  ) {
    return { label: NameSourceType.PIN, value: property.pin.toString() };
  } else if (property.planNumber !== undefined) {
    return { label: NameSourceType.PLAN, value: property.planNumber };
  } else if (property.latitude !== undefined && property.longitude !== undefined) {
    return {
      label: NameSourceType.LOCATION,
      value: compact([property.longitude?.toFixed(6), property.latitude?.toFixed(6)]).join(', '),
    };
  } else if (property.address !== undefined) {
    return {
      label: NameSourceType.ADDRESS,
      value: property.address,
    };
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getPrettyLatLng = (location?: Api_Geometry) =>
  compact([
    location?.coordinate?.x?.toFixed(6) ?? 0,
    location?.coordinate?.y?.toFixed(6) ?? 0,
  ]).join(', ');

export const getFilePropertyName = (
  fileProperty?: Api_PropertyFile,
  skipName: boolean = false,
): PropertyName => {
  if (fileProperty === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  if (
    fileProperty.propertyName !== undefined &&
    fileProperty.propertyName !== '' &&
    skipName === false
  ) {
    return { label: NameSourceType.NAME, value: fileProperty.propertyName };
  } else if (fileProperty.property !== undefined) {
    const property = fileProperty.property;
    return getApiPropertyName(property);
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getApiPropertyName = (property?: Api_Property): PropertyName => {
  if (property === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  let mapProperty: IMapProperty = {
    pin: property.pin?.toString(),
    pid: property.pid?.toString(),
    latitude: property.latitude,
    longitude: property.longitude,
    planNumber: property.planNumber,
    address: property.address !== undefined ? formatApiAddress(property.address) : undefined,
  };
  return getPropertyName(mapProperty);
};
