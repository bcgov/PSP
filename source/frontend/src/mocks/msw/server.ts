import { setupServer } from 'msw/node';

import { handlers } from './handlers';

// This configures a request mocking server with the given request handlers (for jest unit tests).
const server = setupServer(...handlers);

export { server };
