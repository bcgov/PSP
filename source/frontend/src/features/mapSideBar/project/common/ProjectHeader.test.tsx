import { server } from '@/mocks/msw/server';
import { mockProjectGetResponse } from '@/mocks/projects.mock';
import { getUserMock } from '@/mocks/user.mock';
import { ApiGen_Concepts_Project } from '@/models/api/generated/ApiGen_Concepts_Project';
import { prettyFormatUTCDate } from '@/utils';
import { render, RenderOptions } from '@/utils/test-utils';

import ProjectHeader, { IProjectHeaderProps } from './ProjectHeader';
import { http, HttpResponse } from 'msw';

type TestProps = Pick<IProjectHeaderProps, 'project'>;

describe('ProjectHeader component', () => {
  const setup = (props: TestProps, renderOptions: RenderOptions = {}) => {
    const utils = render(<ProjectHeader project={props.project} />, { ...renderOptions });

    return { ...utils };
  };

  let project: ApiGen_Concepts_Project;
  beforeEach(() => {
    project = mockProjectGetResponse();
    server.use(
      http.get('/api/users/info/:userId', () => {
        return HttpResponse.json(getUserMock(), { status: 200 });
      }),
    );
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders as expected', () => {
    const { asFragment } = setup({ project });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders values as expected', () => {
    const { getByText } = setup({ project: project });

    const createDateString = prettyFormatUTCDate(project?.appCreateTimestamp);
    const updateDateString = prettyFormatUTCDate(project?.appLastUpdateTimestamp);

    expect(getByText('771 Project Cool A')).toBeVisible();
    expect(getByText('USER_A')).toBeVisible();
    expect(getByText(new RegExp(createDateString))).toBeVisible();
    expect(getByText('USER_B')).toBeVisible();
    expect(getByText(new RegExp(updateDateString))).toBeVisible();
    expect(getByText('ACTIVE (AC)')).toBeVisible();
  });
});
