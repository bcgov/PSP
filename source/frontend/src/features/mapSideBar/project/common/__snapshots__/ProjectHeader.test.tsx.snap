// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProjectHeader component > renders as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div />
  .c4 {
  font-size: 0.87em;
  line-height: 1.9;
}

.c2 {
  font-weight: bold;
  white-space: nowrap;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.c7 {
  font-family: 'BCSans-Bold';
  font-size: 1.4rem;
  border-radius: 0.4rem;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.2rem 0.5rem;
  margin-left: 0.4rem;
  gap: 0.5rem;
  color: #2e8540;
  background-color: #dff0d8;
}

.c5 {
  margin-top: auto;
  padding-top: 1rem;
  padding-bottom: 1rem;
}

.c0 {
  margin-top: 0.5rem;
  margin-bottom: 1.5rem;
  border-bottom-style: solid;
  border-bottom-color: grey;
  border-bottom-width: 0.1rem;
}

.c3 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  max-width: 60rem;
}

<div
    class="c0 no-gutters row"
  >
    <div
      class="c1 col-7"
    >
      <div
        class="row"
      >
        <div
          class="pr-0 text-left col-3"
        >
          <label
            class="c2"
          >
            Project:
          </label>
        </div>
        <div
          class="pl-1 text-left col-9"
        >
          771 Project Cool A
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="pr-0 text-left col-3"
        >
          <label
            class="c2"
          >
            MoTI region:
          </label>
        </div>
        <div
          class="pl-1 text-left col-9"
        >
          South Coast Region
        </div>
      </div>
    </div>
    <div
      class="col"
    >
      <div
        class="c3"
      >
        <div
          class="no-gutters row"
        >
          <div
            class="text-right col"
          >
            <span
              class="c4"
            >
              <strong>
                Created: 
              </strong>
              Jan 31, 2023 by 
              <span
                class="tooltip-icon"
                data-testid="tooltip-icon-userNameTooltip"
                id="userNameTooltip"
              >
                <strong>
                  USER_B
                </strong>
              </span>
            </span>
          </div>
        </div>
        <div
          class="no-gutters row"
        >
          <div
            class="text-right col"
          >
            <span
              class="c4"
            >
              <strong>
                Updated: 
              </strong>
              Feb 1, 2023 by 
              <span
                class="tooltip-icon"
                data-testid="tooltip-icon-userNameTooltip"
                id="userNameTooltip"
              >
                <strong>
                  USER_A
                </strong>
              </span>
            </span>
          </div>
        </div>
        <div
          class="c5"
        >
          <div
            class="no-gutters justify-content-end align-items-end row"
          >
            <div
              class="col"
            />
            <div
              class="align-self-end d-flex col-auto"
            >
              <span
                class="c4"
              >
                <b />
              </span>
              <div
                class="c6 c7"
              >
                <svg
                  fill="currentColor"
                  height="16"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2z"
                  />
                </svg>
                ACTIVE (AC)
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
